src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp

```diff
@@ -398,82 +409,58 @@ std::vector<DSAInfo> firmwareImage::installedDSA;
-    //         installedDSA.push_back(dsa);
-    //     }
-    //     closedir(dp);
-    // }
-
-    // dp = opendir(FORMATTED_FW_DIR);
-    // if (!dp)
-    //     return installedDSA;
-    // closedir(dp);
-
-    // path formatted_fw_dir(FORMATTED_FW_DIR);
-    // std::vector<std::string> suffix = { XSABIN_FILE_SUFFIX, DSABIN_FILE_SUFFIX};
-
-    // for (std::string t : suffix) {
-
-    //     std::regex e("^" FORMATTED_FW_DIR "/([^/]+)/([^/]+)/([^/]+)/.+\\." + t);
-    //     std::cmatch cm;
-
-    //     for (recursive_directory_iterator iter(formatted_fw_dir, symlink_option::recurse), end;
-    //         iter != end;
-    //         )
-    //     {
-    //         std::string name = iter->path().string();
-    //         std::regex_match(name.c_str(), cm, e);
-    //         if (cm.size() > 0)
-    //         {
-    //             std::string pr_board = cm.str(1);
-    //             std::string pr_family = cm.str(2);
-    //             std::string pr_name = cm.str(3);
-    //             DSAInfo dsa(name, pr_board, pr_family, pr_name);
-    //             installedDSA.push_back(dsa);
-    //             while (iter != end && iter.level() > 2)
-    //                 iter.pop();
-    //         } else if (iter.level() > 4)
-    //             iter.pop();
-    //         else
-    //         {
-    //             dp = opendir(name.c_str());
-	// 	if (!dp)
-    //             {
-    //                 iter.no_push();
-    //             }
-	// 	else
-    //             {
-    //                 closedir(dp);
-    //             }
-    //             ++iter;
-    //         }
-    //     }
-    // }
+    // Obtain installed DSA info.
+    boost::filesystem::path p(FIRMWARE_DIR);
+	if (!boost::filesystem::is_directory(p)) {
+		p = FIRMWARE_WIN_DIR;
+	}
+
+    boost::filesystem::directory_iterator dir_end;
+    for ( boost::filesystem::directory_iterator start(p); start != dir_end; ++start) {
+      std::string filename = start->path().leaf().string();
+      if(filename.find("xsabin")  == std::string::npos && 
+          filename.find("dsabin")  == std::string::npos)
+        continue;
+	  DSAInfo dsa(start->path().string());
+	  installedDSA.push_back(dsa);
+    }
+    if (!installedDSA.empty())
+        return installedDSA;
+
+    // for 2RP
+    p = FORMATTED_FW_DIR;
+	if (!boost::filesystem::is_directory(p))
+        return installedDSA;
+
+    boost::filesystem::path formatted_fw_dir(FORMATTED_FW_DIR);
+    std::vector<std::string> suffix = { XSABIN_FILE_SUFFIX, DSABIN_FILE_SUFFIX};
+
+    for (std::string t : suffix) {
+
+        std::regex e("^" FORMATTED_FW_DIR "/([^/]+)/([^/]+)/([^/]+)/.+\\." + t);
+        std::cmatch cm;
+
+        for (boost::filesystem::recursive_directory_iterator iter(formatted_fw_dir, 
+            boost::filesystem::symlink_option::recurse), recursive_end; iter != recursive_end;) {
+            std::string name = iter->path().string();
+            std::regex_match(name.c_str(), cm, e);
```

Comment:
is c_str() really necessary?

Reply:
yes

Reply:
I don't think so.  I think it is std::smatch sm to stay with std::string.  Please grep code base for std::smatch it's used in different places.

Reply:
Oh ok. I haven't really explored regex so I don't know much about it. I made the change you asked. I'll have to go through the code again when verifying 2RP so I'll take a closer look then :)